share/spack/docker/ci/Dockerfile

```diff
@@ -0,0 +1,74 @@
+ARG BASE
+
+FROM $BASE
+MAINTAINER Spack Maintainers <maintainers@spack.io>
+
+ENV CURRENTLY_BUILDING_DOCKER_IMAGE=1
+
+MASK PUSH
+MASK [[ $DISTRO == centos ]]
+RUN yum update -y                                          \
+ && yum install -y gmp-devel libmpc-devel mpfr-devel which \
+ && rm -rf /var/cache/yum                                  \
+ && yum clean all
+
+MASK [[ $DISTRO == ubuntu ]]
+RUN apt-get -yqq update                                     \
+ && apt-get -yqq install clang g++-5 gcc-5 gfortran-5 unzip \
+ && rm -rf /var/lib/apt/lists/*
+
+
+MASK [[ $DISTRO == centos ]]
+# Download, build and install gcc 5.5.0
+RUN mkdir -p /gcc-bld /opt/gcc/5.5.0                                    \
+ && ( cd /                                                              \
+ &&   curl -OL https://ftp.gnu.org/gnu/gcc/gcc-5.5.0/gcc-5.5.0.tar.xz   \
+ &&   tar -xvf gcc-5.5.0.tar.xz                                         \
+ &&   cd gcc-bld                                                        \
+ &&   ../gcc-5.5.0/configure                                            \
+          --enable-languages=c,c++,fortran                              \
+          --disable-multilib                                            \
+          --prefix=/opt/gcc/5.5.0                                       \
+ &&   make -j\$(nproc)                                                  \
+ &&   make install                                                    ) \
+ && rm -rf /gcc-bld /gcc-5.5.0 /gcc-5.5.0.tar.xz
+
+MASK POP
+
+
+MASK PUSH
+MASK [[ $DISTRO == centos ]]
+RUN spack compiler find /opt/gcc/5.5.0                                       \
+ && FORTRAN='/opt/gcc/5.5.0/bin/gfortran'                                    \
+ && SED_SCRIPT="s|f77: null|f77: \${FORTRAN}|g;s|fc: null|fc: \${FORTRAN}|g" \
+ && sed -i "\${SED_SCRIPT}" ~/.spack/linux/compilers.yaml
+
+MASK [[ $DISTRO == ubuntu ]]
+RUN spack compiler find gcc clang                                            \
+ && FORTRAN='/usr/bin/gfortran'                                              \
+ && SED_SCRIPT="s|f77: null|f77: \${FORTRAN}|g;s|fc: null|fc: \${FORTRAN}|g" \
+ && sed -i "\${SED_SCRIPT}" ~/.spack/linux/compilers.yaml
+MASK POP
+
+
+MASK PUSH
+MASK [[ $DISTRO == centos ]]
+COPY update_rpaths.py /update_rpaths.py
+
+RUN spack python /update_rpaths.py \
+        --prefix /opt/gcc/5.5.0 --rpaths /opt/gcc/5.5.0/lib64
+
+RUN ( spack install      -y llvm@6.0.0%gcc@5.5.0   \
+ ||   spack install -j 1 -y llvm@6.0.0%gcc@5.5.0 ) \
```

Comment:
Is this about trying to handle some issue with the default build parallelism of spack?  Just curious since I didn't run into issues here.  Maybe because I built by hand rather than having Travis do it.

Reply:
Yes.  Specifically, this works around the same issue that requires us to run the beefier M4 AWS instances.  Namely, the high memory watermark that these builds can hit when building with a high parallel factor.  My initial attempts to build on my macbook crashed the compiler because of memory exhaustion on the docker hypervisor (docker for mac uses a hypervisor running a minimal Linux OS, and the amount of memory dedicated to the hypervisor has to be set up-front).

My workaround here boils down to trying to build with default parallel factor, since hey, you very well might have a sufficiently beefy system to handle it, and just try again sequentially if the first try fails.  Since Spack picks up incomplete builds from where they were last left off, I see this as having the best of both worlds: The speed of building in parallel (up to the point where you hit system limits), and the safety and reliability of building sequentially.

We should probably do something similar in CI.

In fact, if there were some way to reliably detect that the compiler crashed due to OOM, I would even suggest that an automatic "retry with parallel back-off" feature be added directly into Spack.

Reply:
Hmm, good suggestion.  I'll give it some thought.